File: src/org/cerion/symcalc/expression/function/list/Reverse.java
package org.cerion.symcalc.expression.function.list;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.ListExpr;

public class Reverse extends FunctionExpr {

	public Reverse(Expr... e) {
		super(FunctionType.REVERSE,e);
	}
	
	@Override
	protected Expr evaluate() {
		
		if(get(0).isList())
		{
			ListExpr L = (ListExpr)get(0);
			ListExpr result = new ListExpr();
			
			for(int i = L.size(); i > 0; i--)
				result.add(L.get(i-1));
			
			return result;
		}
	
		return this;
	}
}

File: src/org/cerion/symcalc/expression/function/list/Reverse.kt
package org.cerion.symcalc.expression.function.list

import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.FunctionExpr
import org.cerion.symcalc.expression.ListExpr

class Reverse(vararg e: Expr) : FunctionExpr(FunctionExpr.FunctionType.REVERSE, *e) {

    override fun evaluate(): Expr {
        if (get(0).isList) {
            val list = get(0).asList()
            val result = ListExpr()

            for (i in list.size() downTo 1)
                result.add(list[i - 1])

            return result
        }

        return this
    }
}